app/javascript/mastodon/service_worker/web_push_notifications.js

```diff
@@ -40,57 +36,82 @@ const notify = options =>
     return self.registration.showNotification(options.title, options);
   });
 
-const handlePush = (event) => {
-  const options = event.data.json();
-
-  options.body      = options.data.nsfw || options.data.content;
-  options.dir       = options.data.dir;
-  options.image     = options.image || undefined; // Null results in a network request (404)
-  options.timestamp = options.timestamp && new Date(options.timestamp);
-
-  const expandAction = options.data.actions.find(action => action.todo === 'expand');
+const fetchFromApi = (path, method, accessToken) => {
+  const url = (new URL(path, self.location)).href;
 
-  if (expandAction) {
-    options.actions          = [expandAction];
-    options.hiddenActions    = options.data.actions.filter(action => action !== expandAction);
-    options.data.hiddenImage = options.image;
-    options.image            = undefined;
-  } else {
-    options.actions = options.data.actions;
-  }
+  return fetch(url, {
+    headers: {
+      'Authorization': `Bearer ${accessToken}`,
+      'Content-Type': 'application/json',
+    },
 
-  event.waitUntil(notify(options));
+    method: method,
+    credentials: 'include',
+  });
 };
 
-const cloneNotification = (notification) => {
-  const clone = {  };
+const formatMessage = (messageId, locale, values = {}) =>
+  (new IntlMessageFormat(locales[locale][messageId], locale)).format(values);
 
-  for(var k in notification) {
-    clone[k] = notification[k];
-  }
+const handlePush = (event) => {
+  const { access_token, notification_id, preferred_locale, title, body, icon } = event.data.json();
+
+  // Placeholder until more information can be loaded
+  event.waitUntil(notify({
```

Comment:
@Gargron I don't think you can call `event.waitUntil` multiple times (or what is the behavior if you do). Something like this might be better suited:

```js
event.waitUntil(
    notify({ ... })
        .then(() => fetchFromApi(...))
        .then(notification => notificationToOptions(notification))
        .then(options => notify(options);
);
```

This way `event.waitUntil` is called only once.

Reply:
what is the reason we can't use async/await in service workers again?

Reply:
The waitUntil() method must be initially called within the event callback, but after that it can be called multiple times, until all the promises passed to it settle.

Reply:
I read the above sentence from the docs multiple times but was still confused by it.

Reply:
i think the intention is "waitUntil must be called at least once but can be called possibly more times"